x/oracle/keeper/historic_price.go

```diff
@@ -0,0 +1,224 @@
+package keeper
+
+import (
+	"sort"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/umee-network/umee/v3/x/oracle/types"
+)
+
+// median returns the median of a list of historic prices.
+func median(prices []types.HistoricPrice) sdk.Dec {
+	lenPrices := len(prices)
+	if lenPrices == 0 {
+		return sdk.ZeroDec()
+	}
+
+	sort.Slice(prices, func(i, j int) bool {
+		return prices[i].ExchangeRate.BigInt().
+			Cmp(prices[j].ExchangeRate.BigInt()) > 0
+	})
+
+	if lenPrices%2 == 0 {
+		return prices[lenPrices/2-1].ExchangeRate.
+			Add(prices[lenPrices/2].ExchangeRate).
+			QuoInt64(2)
+	}
+	return prices[lenPrices/2].ExchangeRate
+}
+
+// medianDeviation returns the standard deviation around the
+// median of a list of prices.
+// medianDeviation = ∑((price - median)^2 / len(prices))
+func medianDeviation(median sdk.Dec, prices []types.HistoricPrice) sdk.Dec {
+	lenPrices := len(prices)
+	medianDeviation := sdk.ZeroDec()
+
+	for _, price := range prices {
+		medianDeviation = medianDeviation.Add(price.ExchangeRate.
+			Sub(median).Abs().Power(2).
+			QuoInt64(int64(lenPrices)))
+	}
+
+	return medianDeviation
+}
+
+// GetMedian returns a given denom's median price in the last prune
+// period since a given block.
+func (k Keeper) GetMedian(
+	ctx sdk.Context,
+	denom string,
+	blockNum uint64,
+) (sdk.Dec, error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.GetMedianKey(denom, blockNum))
+	if bz == nil {
+		return sdk.ZeroDec(), sdkerrors.Wrap(types.ErrUnknownDenom, denom)
```

Comment:
It might make more sense to do a new error message here, since initially medians will not exist until the first `Median Period` passes